src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -548,12 +548,24 @@ public void setUserProperties(JSONObject userProperties) {
     }
 
     public void setUserProperties(final JSONObject userProperties, final boolean replace) {
-        if (replace || this.userProperties == null) {
-            this.userProperties = userProperties;
+        if (userProperties == null) {
+            if (replace) {
+                this.userProperties = null;
+            }
             return;
         }
 
-        if (userProperties == null) {
+        if (replace || this.userProperties == null) {
+            // Only create copy on main thread if necessary
```

Comment:
I was originally thinking about that too, but I was worried about the capture in the run. getInstance() grabs the wrong instance, but looks like I can capture "this" for the run.